Now playing doesn't stay in sync with MPD / moode / large collection

While trying out euphonica 0.97.0 I noticed that the "Now playing" can be desyncronised with my MPD player. 

Here is a screenshot side by side with https://github.com/yktoo/ymuse, the left side is correct as to which track is currently playing. 

<img width="1901" height="1039" alt="Image" src="https://github.com/user-attachments/assets/3a6e3d8e-aaa9-470c-bde3-89be4724b84c" />

One possible problem here is : is it possible I put my laptop to sleep and on resume the MPD connection did not resume ? 

Actually, controls don't respond on euphonica although it states it is "Connected"

Maybe using async io can help instead of migrating logic into threads. Threads often introduce complexity that shouldn't be needed on an app like this. It would still be a bit of work but usually that way the result is more maintainable in the end.

On my debian stable, unfortunately I cannot get #194 to build
```
  Compiling nohash-hasher v0.2.0
   Compiling euphonica v0.97.2 (/home/christof/Programmierung/euphonica)
warning: unused import: `Result as ZResult`
  --> src/client/wrapper.rs:15:56
   |
15 | use zbus::{Connection as ZConnection, Proxy as ZProxy, Result as ZResult};
   |                                                        ^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `std::num::NonZero`
   --> src/player/controller.rs:167:9
    |
167 |     use std::num::NonZero;
    |         ^^^^^^^^^^^^^^^^^

warning: unused import: `ImageReader`
 --> src/utils.rs:6:35
  |
6 | use image::{imageops::FilterType, ImageReader, DynamicImage, RgbImage};
  |                                   ^^^^^^^^^^^

warning: unused import: `io::Cursor`
  --> src/utils.rs:11:11
   |
11 | use std::{io::Cursor, sync::RwLock};
   |           ^^^^^^^^^^

error[E0658]: use of unstable library feature `inherent_str_constructors`
  --> src/meta_providers/base.rs:61:40
   |
61 |                         if let Ok(s) = str::from_utf8(&bytes) {
   |                                        ^^^^^^^^^^^^^^
   |
   = note: see issue #131114 <https://github.com/rust-lang/rust/issues/131114> for more information

For more information about this error, try `rustc --explain E0658`.
warning: `euphonica` (bin "euphonica") generated 4 warnings
error: could not compile `euphonica` (bin "euphonica") due to 1 previous error; 4 warnings emitted
FAILED: src/euphonica 
/usr/bin/env CARGO_HOME=/home/christof/Programmierung/euphonica/build/cargo-home /home/christof/.cargo/bin/cargo build --manifest-path /home/christof/Programmierung/euphonica/Cargo.toml --target-dir /home/christof/Programmierung/euphonica/build/src --release && cp src/release/euphonica src/euphonica
ninja: build stopped: subcommand failed.
Could not rebuild /home/christof/Programmierung/euphonica/build

```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -52,6 +52,7 @@ resolve-path = "0.1.0"
 lru = "0.16.0"
 nohash-hasher = "0.2.0"
 libsecret = "0.7.0"
+zbus = { version = "5.12.0", features = ["tokio"], default-features = false}
 
 [dependencies.gtk]
 package = "gtk4"

--- a/src/client/wrapper.rs
+++ b/src/client/wrapper.rs
@@ -12,6 +12,7 @@ use mpd::{
     song::Id,
     Channel, EditAction, Idle, Output, SaveMode, Subsystem,
 };
+use zbus::{Connection as ZConnection, Proxy as ZProxy, Result as ZResult};
 
 use std::net::TcpStream;
 use std::os::unix::net::UnixStream;
@@ -387,6 +388,40 @@ impl MpdWrapper {
                     glib::timeout_future_seconds(ping_interval).await;
                 }
             }));
+
+        // A new loop to watch for system suspend/wake actions. We proactively disconnect before suspend
+        // and connect again upon wake to avoid freezing upon a timed-out connection.
+        // let (suspend_sender, suspend_receiver): (Sender<bool>, Receiver<bool>) = async_channel::unbounded();
+        let main_sender = self.main_sender.clone();
+        utils::tokio_runtime().spawn(async move {
+            let connection = ZConnection::system().await.unwrap();
+
+            // Create a proxy for systemd-logind
+            let proxy = ZProxy::new(
+                &connection,
+                "org.freedesktop.login1",      // The service name
+                "/org/freedesktop/login1",     // The object path
+                "org.freedesktop.login1.Manager", // The interface
+            ).await.unwrap();
+
+            // Get a stream of the "PrepareForSleep" signal
+            use futures::prelude::*;
+            let mut signal_stream = std::pin::pin!(
+                proxy.receive_signal("PrepareForSleep").await.unwrap()
+            );
+            // Loop forever, processing signals
+            while let Some(signal) = signal_stream.next().await {
+                if let Ok(going_to_sleep) = signal.body().deserialize::<bool>() {
+                    if going_to_sleep {
+                        println!("System is preparing to suspend. Disconnecting...");
+                        main_sender.send(AsyncClientMessage::Disconnect).await;
+                    } else {
+                        println!("System has woken up. Reconnecting...");
+                        main_sender.send(AsyncClientMessage::Connect).await;
+                    }
+                }
+            }
+        });
     }
 
     async fn respond(&self, request: AsyncClientMessage, recently_connected: bool) -> glib::ControlFlow {

